### DISCUSSION TITLE
Multiple API Request in getInitialProps and getServerSideProps slow Down the performance - Next.js

### DISCUSSION THREAD

**Discussion by faisal78676** [vercel/next.js#27075] [Help]
Please help to fix this issue

I am building a dynamic website for displaying live match scores and stats everything is fine.

Before deploying on the server we test the performance and stress test on our staging environment.

**The Mobile score is 50** and **the Desktop score is 80** on the performance test.   
Lighthouse reporting the **google tag manager** and **DFP** takes time to execute and Initial server response time takes time

My question is how to reduce server initial response time and javascript execution time. How I improve performance that below are details what I did in the code

1) How to load third-party scripts without blocking time like **Google Tag Manager** and **DFP scripts**. I have added **GA** code on **_document.js** and for DFP using **react-dfp** package 
but it's slow down the performance

2) **header**, **header2** and **Right bar** and **post list** components should **Refresh** after the 1-Min of interval for that I am using **useEffect()** hook

4) Fetching **Header, Footer, Rightbar, LeftBar** component data from API inside the **getInitialProps** with the help of **Axios**. These components are common for all the pages
3) Fetching slider and post list Component data inside **getServerSideProps** with the help of **Axios** package

**Header, Header 2, Left, Right, Footer**

Above component imported using `dynamic () import` function with `ssr:false`

Imported below component normally

**Slider**, **Post list**

Below the sample code.

     ********** _app.js **********
   
     MyApp.getInitialProps = async (appContext) => {    
           const appProps = await App.getInitialProps(appContext);
          //Common Component
           const header = await axios(header_api)
           const header2 = await axios(header2_api)   
           const left = await axios(left_api)
           const right = await axios(right_api)
           const footer = await axios(footer_api)
        
           return {        
               appProps:{
                   header,
                   header2,
                   left,
                   right,
                   footer
               } 
            }
         }

```

    ************* Home.js ***********
    export async function getServerSideProps(context){    
       const slider = await axios(slider_api)
       const postlist = await axios(postlist_api)    
       return {
          props :{ 
                  slider,
                  postlist
          }    
     }

```
```
*********** _document.js ***************
class MyDocument extends Document {  
  
  render() {
    console.log('__document.js');    
    return (
      <Html lang="en">
      <Head>        
                ******* other third party scripts *** 
                <script async dangerouslySetInnerHTML={{ __html: `(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);})(window,document,'script','dataLayer','XXXXXXXXX');`}}/>
      </Head>
      <body id="body-wrap">
      <noscript dangerouslySetInnerHTML={{ __html: `<iframe src="https://www.googletagmanager.com/ns.html?id=XXXXXXX" height="0" width="0" style="display:none;visibility:hidden"></iframe>`}}></noscript>
        <Main />        
        <NextScript />        
      </body>            
    </Html>
    )
  }
}
```

**Comment by UncleClapton** [comment]
Hi there, If you're still experiencing this issue I think I can provide some initial help, but there may be even faster alternatives that could be suggested if I had full context.

Right now you're making each request sequentially. One must conclude before the other may begin. This is likely the main cause of slow response times for your app.

```js
MyApp.getInitialProps = async (appContext) => {
  const header = await axios(header_api) // Runs..
  const header2 = await axios(header2_api) // must wait for header
  const left = await axios(left_api) // must wait for header and header2
  // and so on...
}
```

Instead you could run these in parallel by using `Promise.all` to queue them all up at once. This comes with it's own trade off, but this would be a significant improvement over your current code. The downside here is that this solution does not scale well. Browsers have a limit to how many connections you can hold at a time.

```js
MyApp.getInitialProps = async (appContext) => {
	const [
		header, 
		header2, 
		left, 
		right, 
		footer
	] = await Promise.all([
		axios(header_api)
		axios(header2_api)
		axios(left_api)
		axios(right_api)
		axios(footer_api)
	]);
	// ... Rest of your logic
}
```

The same applies to your `getServerSideProps` example

```js
export async function getServerSideProps (context) {    
	const [slider, postlist] = await Promise.all([axios(slider_api), axios(postlist_api)])
	return {
		props: { 
			slider,
			postlist
		}
	}
}
```


Beyond that, A much more scalable answer to your problem would be reworking your API to respond with all of the data you need for your layout in a single request. Is there a reason you need each Component's data requested separately?
